plugins/out_s3/s3.c

```diff
@@ -1114,11 +1164,28 @@ static int upload_data(struct flb_s3 *ctx, struct s3_file *chunk,
         if (ctx->compression == FLB_AWS_COMPRESS_GZIP) {
             flb_free(payload_buf);
         }
-        m_upload->upload_errors += 1;
         /* re-add chunk to list */
         if (chunk) {
-            s3_store_file_unlock(chunk);
             chunk->failures += 1;
+            if (chunk->failures > ctx->ins->retry_limit) {
+                s3_retry_warn(ctx, (char *) chunk->fsf->meta_buf, m_upload->input_name,
+                              chunk->create_time, FLB_FALSE);
+                s3_store_file_delete(ctx, chunk);
+                /* destory multipart upload if part_number is 1, else set as complete*/
+                if (m_upload->part_number == 1) {
+                    multipart_upload_destroy(m_upload);
```

Comment:
Need to also remove it from the uploads list otherwise we'll have an issue: `mk_list_del(&m_upload->_head);`